### DISCUSSION TITLE
[API Design] Link<?> generic signature

### DISCUSSION THREAD

**Discussion by stefanseifert** [adobe/aem-core-wcm-components#1853] [Ideas]
i currently see no benefit with the concept of having a generic signature like `Link<Page>` or `Link<Asset>`. in theory it makes sense, in practice not.

whenever you consume a link object from link handler or from an model interface, you have no idea on what item the link really points to when you declare your variable, and you can only use `Link<?>` or `Link`. same applies in model interfaces that return links. you only know what's in it after you actually got the link instance. this makes using link objects very clumsy, and you already see code scattered with @SuppressWarnings("rawtypes") to get rid of related warnings.

the only place where you are sure what the link is pointing at is during link building/handling, but this piece of code is inside the link handling and not relevant for the users of the API.

i would propose to drop this generic concept completely.

---

this discussion is taken over from the [mailing list](https://groups.google.com/d/msgid/aem-core-components-dev/AS8PR07MB7238137D48504B293171B7F1B8C89%40AS8PR07MB7238.eurprd07.prod.outlook.com) where it did not gather any attention (except from @ky940819).

**Comment by vladbailescu** [comment]
I think we can drop the generic support, it hasn't proved useful.
